www/js/survey/enketo-survey-services.js

```diff
@@ -0,0 +1,117 @@
+angular.module('emission.enketo-survey.services', ['ionic', 'emission.plugin.logger'])
+.factory('EnketoSurvey', function($window, $http, Logger) {
+  var __form = null;
+  var __session = {};
+  var __form_location = null;
+  var __loaded_form = null;
+  var __loaded_model = null;
+
+  function init(formLocParam, optsParam) {
+    __form = null;
+    __session = {};
+    __form_location = formLocParam;
+    __loaded_form = null;
+    __loaded_model = null;
+
+    const opts = JSON.parse(optsParam);
+    if (opts && opts.session) {
+      __session = opts.session;
+    }
+
+    return $http.get(__form_location)
+    .then(function(form_json) {
+      __loaded_form = form_json.data.form;
+      __loaded_model = form_json.data.model;
+    });
+  }
+
+  function _loadForm(opts = {}) {
+    var formSelector = 'form.or:eq(0)';
+    var data = {
+      // required string of the default instance defined in the XForm
+      modelStr: __loaded_model,
+      // optional string of an existing instance to be edited
+      instanceStr: opts.instanceStr || null,
+      submitted: opts.submitted || false,
+      external: opts.external || [],
+      session: opts.session || {}
+    };
+    __form = new $window.FormModule( formSelector, data, {});
+    var loadErrors = __form.init();
+    return loadErrors;
+  }
+  
+  function _restoreAnswer(answers) {
+    Logger.log("Answers stored locally" + JSON.stringify(answers));
+    var tripAnswer = null;
+    var found = false;
+    answers.forEach(function(answer){
+      Logger.log(__session.trip_properties);
+      Logger.log(answer.trip_properties);
+      if ((__session.trip_properties.start_ts === answer.trip_properties.start_ts) &&
```

Comment:
for the matching, please consider using `DiaryHelper.getUserInputForTrip`
https://github.com/e-mission/e-mission-phone/blob/master/www/js/diary/services.js#L415
to avoid code duplication

Note that it works for draft trips too, although you may not need that functionality.

Reply:
That's awesome, although there are minor changes need to be made on `DiaryHelper.getUserInputForTrip`. But it makes the code cleaner.